### PR TITLE
[agent] LLM-based editing

### PR DESCRIPTION
**Pull request by xingyaoww**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

Hand-crafted editing tools have been causing a lot of problems:
-  we have issues in search/replace editing using Python (i.e., weird escapes causing newline to not work #3650)
- even if search/replace block is re-written to be parsed directly into `EditAction` which should fixed the escaping issue, the search/replace block logic might still be tricky for LLM to handle, which is evident in [this benchmark ('whole' file output performance > 'diff' output)](https://aider.chat/docs/benchmarks.html).

Cursor had an initial blog post (now removed, [original link](https://www.cursor.com/blog/instant-apply)) that talks about how "instant apply" works (#1854). In essence, the idea is 
1. Use the stronger LLM (e.g., claude 3.5) to read the original file, and propose a draft edit (e.g., omit certain locations that are not changed).
2. Use another LLM (e.g., cursor trained a specialized model, this PR uses smaller LLMs like `GPT-4o-mini` and `haiku`) to read the "original file" and "the draft edit" and re-generate the new output file (i.e., edited file).

The benefit here is that we don't need to manually implement complex logic of file editing tools to try to get the indentation right, search/replace logic clear to the agent, etc. The downside is that the "draft apply" process might introduce a bit more non-deterministic, which calls for more extensive evaluation on specifically the "file draft apply" task.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

Updates 10/15/2024:

> Sharing some updates, I think we are close to having the editing PR merged! It seems the upper bound of the new editing is higher than our current main:
> 
> * **Performance-wise**: On editing PR iteration v2.4 - We got to 77/300 (25.67%) w/ sonnet by increasing the max iteration to 100 (default of our app i believe) with avg 41 turns. The main baseline got 63 / 300 (21.00%) using the same 100 iterations requiring avg 43 turns (there were some performance degradation on our main branch since `24 sep` eval - @neubig is currently helping me locating the issue now).
> * **Cost-wise**:
>   
>   * A bump to 100 max turns increases costs to about 1$ per instance (about where we were when there was no prompt caching), whereas we previously measured sonnet costs about $0.3 per issue (but there was some bug in the metric calculation so this could be under-estimate)
>   * We CAN use the default agent model as an editing model without incurring cost increases.
>     
>     * On editing PR iteration v2.4
>       
>       * Avg. agent cost per instance: 0.92 USD
>       * Avg. editor cost per instance: 0.12 USD
>       * Avg. total cost per instance: 1.04 USD
>     * On main baseline (where no LLM-based editing is used):
>       
>       * Avg. num of turns per instance: 43.02
>       * Avg. agent cost per instance: 1.27 USD
>       * Avg. editor cost per instance: 0.00 USD
>       * Avg. total cost per instance: 1.27 USD
> * Increase the file-context window from 100 to 300 lines doesn't work
>   
>   * anthropic/claude-3-5-sonnet-20240620 WITH 300 lines file ctx: 63/300 = 21%
>   * anthropic/claude-3-5-sonnet-20240620: 77/300 = 25.7%
> * New editing PR seems to suffer less from exceeding 200k context window of Claude
>   
>   * for editing v2.4: 1/300
>   * for main branch baseline: 12/300
> 
> ALL the experiments above are performed using anthropic/claude-3-5-sonnet-20240620, with one run each (can't afford to run it multiple times now we increase the turns to 100 - back at square 1 where each run costs 300 bucks).
> 
> I'll gradually move pieces from the editing PR into separate PR to make it easier for code review & fixing test cases

---
**Link of any specific issues this addresses**

### CODE REVIEW

**Comment by tobitege** [comment]
Great idea, exciting! 🚀

File: agenthub/codeact_agent/action_parser.py
Comment (neubig) [inline agenthub/codeact_agent/action_parser.py:54]:
```suggestion
        if '<file_edit' in action and '</file_edit>' not in action:
            action += '</file_edit>'
```
Don't need to detect file_edit for every language.

File: agenthub/codeact_agent/action_parser.py
Comment (neubig) [inline agenthub/codeact_agent/action_parser.py:220]:
```suggestion
        if '<file_edit' not in action_str:
            return False
            
        self.file_edit_match = re.search(
            r'<file_edit\s+path=(["\']?)(.*?)\1>(.*?)</file_edit>',
            action_str,
            re.DOTALL,
        )
        path = None if self.file_edit_match is None else self.file_edit_match.group(2)
        if not path:
            raise ActionParseError(
                error='FileEditAction detected but no `path` specified. You should specify the path of the file to edit by setting `path="<path>"` in <file_edit> tag. For example: <file_edit path="path/to/file.txt">...</file_edit>'
            )
        
        return True
```

It seems this could be simplified.

File: agenthub/codeact_agent/codeact_agent.py
Comment (neubig) [inline agenthub/codeact_agent/codeact_agent.py:101]:
```suggestion
```
Delete before merging.

File: evaluation/swe_bench/README.md
Comment (neubig) [inline evaluation/swe_bench/README.md:160]: The changes in this file seem orthogonal to this PR?

File: evaluation/aider_bench/run_infer.py
Comment (neubig) [inline evaluation/aider_bench/run_infer.py:74]: This is a bit convoluted and potentially error-prone. I feel there should be a better way to do this (e.g. including `draft_editor` in `metadata`).

File: evaluation/aider_bench/scripts/summarize_results.py
Comment (neubig) [inline evaluation/aider_bench/scripts/summarize_results.py:1]: I didn't fully follow the changes in this file, and they seem orthogonal to the main point of this commit?

**Review by neubig — COMMENTED** [review]
Hey, I took an initial look. Overall this looks OK, but it seems like there might be quite a few unrelated changes (such as adding `swe_bench/eval_infer.py`). Could I take a look after it's simplified down to be a smaller set of coherent changes?

File: evaluation/swe_bench/README.md
Comment (xingyaoww) [reply evaluation/swe_bench/README.md:160]: oh yeah these are due to the fact that this PR hasn't synced with the latest main - can be solved with a later merge.

File: evaluation/aider_bench/scripts/summarize_results.py
Comment (xingyaoww) [reply evaluation/aider_bench/scripts/summarize_results.py:1]: yep - i can break this into a separate PR (or just remove this changes of this file completely)

File: evaluation/aider_bench/run_infer.py
Comment (xingyaoww) [reply evaluation/aider_bench/run_infer.py:74]: yeah.. i've also been thinking about this in the back of my mind - we probably also need to figure out a way to set this in the UI as well, so it is probably a larger changes..

Comment (xingyaoww) [reply evaluation/aider_bench/run_infer.py]: I move this to a field in `LLMConfig` in newer commits - hopefully will make things clearer.

File: agenthub/codeact_agent/codeact_agent.py
Comment (enyst) [inline agenthub/codeact_agent/codeact_agent.py:231]: Just a heads-up, I merged the revert of this Exception. I really think it will prove simpler to work with, in practice.

File: agenthub/codeact_agent/system_prompt.j2
Comment (enyst) [inline agenthub/codeact_agent/system_prompt.j2:22]: We could even go with MUST SPECIFY, why not 😄

Comment (xingyaoww) [reply agenthub/codeact_agent/system_prompt.j2:22]: I REPEAT! YOU MUST ... 😄

File: agenthub/codeact_agent/codeact_agent.py
Comment (enyst) [inline agenthub/codeact_agent/codeact_agent.py:225]: Just a thought here, we already have a little mechanism to handle parsing errors in the LLM response, maybe we can use it? We catch this kind of errors [here](https://github.com/All-Hands-AI/OpenHands/blob/1354675ce35e28840a7c58f517380ee553e25007/openhands/controller/agent_controller.py#L425) and then `send_error` adds the ErrorObs to the stream, the LLM will receive it next time.

It also shows it to the user, maybe we don't want that part here, then we could add another `except` to only handle the obs.

The old exceptions caught at that line really are parsing errors, as they were used for json agents, stuff like missing keys in the response, or the wrong values or types in the response for a particular action. Sounds like this!

Comment (xingyaoww) [reply agenthub/codeact_agent/codeact_agent.py:225]: Good point! This should ideally make things easier - I'll revise

File: openhands/utils/chunk_localizer.py
Comment (enyst) [inline openhands/utils/chunk_localizer.py:97]: I wonder, what if we could set a threshold, to filter out too unlikely chunks? Like, let's say k=3, max_chunk_size=20, and we have one result 15/20, and the others 2/20 and 2/20. Does it make sense to send in the context those with 0.1 score? If the distance is too high between scores, maybe we can avoid to send the "worse" ones... if that makes any sense! WDYT?

**Comment by enyst** [comment]
This is an awesome experiment! I read it tonight in full - finally 😊

Thank you for doing this!

File: openhands/runtime/utils/edit.py
Comment (enyst) [inline openhands/runtime/utils/edit.py:92]: Curiously, Sonnet 3.5 has 8k. 🤔 Not in scope of this PR, maybe we can try sometime full file edit when possible, when the file is up to 2-3k...

Comment (xingyaoww) [reply openhands/runtime/utils/edit.py:92]: 300 lines is roughly around > 2k tokens in my testing & most cheap LLMs can do 4k at least. But qualitatively I find that producing more lines seem to have worse quality. We have some one working on an eval benchmark for this :D - once that's done, we can look at qualitative number and determine the best value for this :)

File: openhands/utils/chunk_localizer.py
Comment (xingyaoww) [reply openhands/utils/chunk_localizer.py:97]: Yep! That is also an option. Right now, I'm just doing YOLO runs without any design justification but instincts 😅 - Once everything is in place (and kinda working), we can start tune these knobs!

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (li-boxuan) [inline openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:432]: This doesn't work in many cases; documented in https://github.com/All-Hands-AI/OpenHands/pull/3649

**Comment by li-boxuan** [comment]
How do we plan to test this PR in integration testing?

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (xingyaoww) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:432]: Yeah - so ideally, if we merge this in, we will get rid of this function altogether AND use this directly:

https://github.com/All-Hands-AI/OpenHands/blob/3ee68b341ee73810999b09a1aaf3e31ec3fc86d4/openhands/linter/linter.py#L38-L66

File: openhands/linter/linter.py
Comment (li-boxuan) [inline openhands/linter/linter.py:64]: Is `new_lineno` guaranteed to be non-conflicting with old line numbers (before edit)?

File: openhands/runtime/plugins/agent_skills/file_ops/file_ops.py
Comment (li-boxuan) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:432]: Cool! You might want to consider reuse some of the test cases here: https://github.com/All-Hands-AI/OpenHands/pull/3649/files#diff-f1e8c3680edd5112543092c9adf55f4cbf79716880a05118b473baf63d7900c8

Comment (tobitege) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:432]: How does linting a diff actually work? Wouldn't it give you potentially loads of errors about methods or variables not defined or something? 🤔

Comment (li-boxuan) [reply openhands/runtime/plugins/agent_skills/file_ops/file_ops.py:432]: My best guess is they only lint the new file, but the new file is organized in a way such that the diff part has their special line numbers that are not to be confused with original file's line numbers... but maybe it's not how it works lol.

**Comment by tobitege** [comment]
Fyi, I think the `--max-chars` in almost all `run_infer.sh` scripts can be removed?

**Comment by xingyaoww** [comment]
@tobitege yep! I did that in https://github.com/All-Hands-AI/OpenHands/pull/4396

**Comment by xingyaoww** [comment]
Sharing some updates, I think we are close to having the editing PR merged! It seems the upper bound of the new editing is higher than our current main:

- **Performance-wise**: On editing PR iteration v2.4 - We got to 77/300 (25.67%) w/ sonnet by increasing the max iteration to 100 (default of our app i believe) with avg 41 turns. The main baseline got 63 / 300 (21.00%) using the same 100 iterations requiring avg 43 turns (there were some performance degradation on our main branch since `24 sep` eval - @neubig is currently helping me locating the issue now).
- **Cost-wise**:
    - A bump to 100 max turns increases costs to about 1$ per instance (about where we were when there was no prompt caching), whereas we previously measured sonnet costs about $0.3 per issue (but there was some bug in the metric calculation so this could be under-estimate)
    - We CAN use the default agent model as an editing model without incurring cost increases.
        - On editing PR iteration v2.4
            - Avg. agent cost per instance: 0.92 USD
            - Avg. editor cost per instance: 0.12 USD
            - Avg. total cost per instance: 1.04 USD
        - On main baseline (where no LLM-based editing is used):
            - Avg. num of turns per instance: 43.02
            - Avg. agent cost per instance: 1.27 USD
            - Avg. editor cost per instance: 0.00 USD
            - Avg. total cost per instance: 1.27 USD
- Increase the file-context window from 100 to 300 lines doesn't work
    - anthropic/claude-3-5-sonnet-20240620 WITH 300 lines file ctx: 63/300 = 21%
    - anthropic/claude-3-5-sonnet-20240620: 77/300 = 25.7%
- New editing PR seems to suffer less from exceeding 200k context window of Claude
    - for editing v2.4: 1/300
    - for main branch baseline: 12/300

ALL the experiments above are performed using anthropic/claude-3-5-sonnet-20240620, with one run each (can't afford to run it multiple times now we increase the turns to 100 - back at square 1 where each run costs 300 bucks).

I'll gradually move pieces from the editing PR into separate PR to make it easier for code review & fixing test cases

File: openhands/linter/linter.py
Comment (xingyaoww) [reply openhands/linter/linter.py:64]: What do you mean by "non-conflicting"? the `new_lineno` is the changed lines in the new file - which should only relevant in the new version of the file?

**Comment by xingyaoww** [comment]
> How do we plan to test this PR in integration testing?
@li-boxuan This is a very good Q... 😓

File: openhands/linter/linter.py
Comment (xingyaoww) [reply openhands/linter/linter.py:64]: I move the linter related changes to a separate PR - will try to add more testcases from your branch :)
https://github.com/All-Hands-AI/OpenHands/pull/4409

**Review by enyst — APPROVED** [review]

**Comment by xingyaoww** [comment]
@li-boxuan 

After discussing with @rbren, we are considering removing the integration test from our CI since the editing now requires LLM to run, which would cause non-deterministic issues & further slow down our CI.

We are thinking about converting the existing integration tests into evaluation instances and then adding nightly runs of Github actions that run a bunch of evaluation benchmarks using the cheap deepseek-v2.5 model, including the integration tests & SWE-Bench. This might help us track performance degradation better 🤔

**Comment by enyst** [comment]
The nightly evals sound like a great idea! Some of the eval instances are quite difficult or long tasks, and running even a few of those can already show curious things.

The results are interesting. I will come back on them and some details of this PR.

**Comment by li-boxuan** [comment]
> After discussing with @rbren, we are considering removing the integration test from our CI since the editing now requires LLM to run, which would cause non-deterministic issues & further slow down our CI.
We are thinking about converting the existing integration tests into evaluation instances and then adding nightly runs of Github actions that run a bunch of evaluation benchmarks using the cheap deepseek-v2.5 model, including the integration tests & SWE-Bench. This might help us track performance degradation better 

I agree and I kinda foresaw this would eventually happen. Integration testing with mocked LLM inputs & responses were useful at the early stage of this project, when refactoring happened frequently and tons of commits checked into the repo everyday without being well-tested. And more importantly, back then it was unrealistic to run benchmarks daily due to cost and infra limitations.

Now it's time for the new chapter and I'd love to see how daily evaluation goes 👍

File: openhands/linter/linter.py
Comment (li-boxuan) [reply openhands/linter/linter.py:64]: > What do you mean by "non-conflicting"?

Never mind, I thought you also linted twice in your approach. Now it seems you only lint once. Anyways let's discuss in your new PR.

**Comment by xingyaoww** [comment]
Now integration tests is removed by https://github.com/All-Hands-AI/OpenHands/pull/4447 to clean-up the diff - this PR is ready for review!

**Comment by xingyaoww** [comment]
(there seems to be something wrong with linter that dropped the perf down to 17% - investigating

**Review by neubig — COMMENTED** [review]
Please re-request when this is ready to merge and I'll take a look  if someone else doesn't do it first!

File: openhands/agenthub/codeact_agent/codeact_agent.py
Comment (neubig) [inline openhands/agenthub/codeact_agent/codeact_agent.py:39]: CodeAct 2.0 :)

File: openhands/core/config/llm_config.py
Comment (neubig) [inline openhands/core/config/llm_config.py:43]:
```suggestion
        draft_editor: A more efficient LLM to use for file editing. Introduced in [PR 3985](https://github.com/All-Hands-AI/OpenHands/pull/3985).
```

File: openhands/core/config/utils.py
Comment (neubig) [inline openhands/core/config/utils.py:36]: It seems like it would be better for this to be in the `LLMConfig` class itself?

File: openhands/llm/metrics.py
Comment (neubig) [inline openhands/llm/metrics.py:18]: Do we want to have a default here? We could also make it required to make sure that it gets recorded.

File: openhands/runtime/client/client.py
Comment (neubig) [inline openhands/runtime/client/client.py:300]:
```suggestion
        # ignore the hostname AND use 'openhands-workspace'
```

File: openhands/runtime/runtime.py
Comment (neubig) [inline openhands/runtime/runtime.py:46]: It's not clear to me why we need to have a MixIn if it's mixed in to the base class? Couldn't we just add it to the base class?

File: openhands/runtime/utils/edit.py
Comment (neubig) [inline openhands/runtime/utils/edit.py:111]: As we discussed face-to-face, the cost may not be so high. Do we just want to remove this warning?

**Review by neubig — COMMENTED** [review]
This was a bit big for me to check super-carefully, but I trust the tests/evals to figure out any major issues. I left a few comments on the things that I noticed.

File: evaluation/swe_bench/run_infer.py
Comment (xingyaoww) [reply evaluation/swe_bench/run_infer.py:87]: Under the same setting, it got from 19% to 20% using sonnet, so probably no significant changes there. Should we revert it then? (I do see some cases where the agent tries to use default Py interpreter though)

File: evaluation/swe_bench/eval_infer.py
Comment (xingyaoww) [reply evaluation/swe_bench/eval_infer.py:77]: ops, i think we can remove this now bc it no longer throws critical error when draft_editor is not set.

File: openhands/runtime/runtime.py
Comment (xingyaoww) [reply openhands/runtime/runtime.py:46]: The main reason is that i kinda want these editing logic to stay in a separate file until they are finalized so the base class is a little cleaner

**Review by xingyaoww — COMMENTED** [review]
Most of stuff should be resolved now!

File: openhands/runtime/utils/edit.py
Comment (xingyaoww) [reply openhands/runtime/utils/edit.py]: removed!

File: evaluation/swe_bench/eval_infer.py
Comment (xingyaoww) [reply evaluation/swe_bench/eval_infer.py]: removed

File: openhands/core/config/utils.py
Comment (xingyaoww) [reply openhands/core/config/utils.py]: good point, refactor it to a class method.

File: openhands/llm/metrics.py
Comment (xingyaoww) [reply openhands/llm/metrics.py:18]: ok.. i think we should leave a default value because the `State` may need to construct a copy of it.

**Review by neubig — APPROVED** [review]
Sounds good!

**Pull request merged**
